tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,213 @@
+
+    uvwasi_size_t nread_;
+    ret = uvwasi_fd_read(&state, static_cast<uvwasi_fd_t>(fd), iovs.data(),
+        static_cast<uvwasi_size_t>(iovs.size()), &nread_);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), nread, nread_);
+
+    return ret;
+}
+
+fizzy::ExecutionResult wasi_fd_prestat_get(
+    fizzy::Instance& instance, fizzy::span<const uint64_t> args, int)
+{
+    const auto fd = static_cast<uint32_t>(args[0]);
+    const auto prestat_ptr = static_cast<uint32_t>(args[1]);
+
+    uvwasi_prestat_t buf;
+    uvwasi_errno_t ret = uvwasi_fd_prestat_get(&state, fd, &buf);
+
+    uvwasi_serdes_write_prestat_t(instance.memory->data(), prestat_ptr, &buf);
+
+    return ret;
+}
+
+fizzy::ExecutionResult wasi_environ_sizes_get(
+    fizzy::Instance& instance, fizzy::span<const uint64_t> args, int)
+{
+    const auto environc = static_cast<uint32_t>(args[0]);
+    const auto environ_buf_size = static_cast<uint32_t>(args[1]);
+    // TODO: implement properly (only returns 0 now)
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), environc, 0);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), environ_buf_size, 0);
+    return UVWASI_ESUCCESS;
+}
+
+const std::vector<fizzy::ImportedFunction> wasi_functions = {
+    {"wasi_snapshot_preview1", "proc_exit", {fizzy::ValType::i32}, std::nullopt, wasi_proc_exit},
+    {"wasi_snapshot_preview1", "fd_read",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_read},
+    {"wasi_snapshot_preview1", "fd_write",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_write},
+    {"wasi_snapshot_preview1", "fd_prestat_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_prestat_get},
+    {"wasi_snapshot_preview1", "fd_prestat_dir_name",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32}, fizzy::ValType::i32,
+        wasi_return_enosys},
+    {"wasi_snapshot_preview1", "environ_sizes_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_environ_sizes_get},
+    {"wasi_snapshot_preview1", "environ_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_return_enosys},
+};
+
+bool run(int argc, const char** argv)
+{
+    uvwasi_options_t options = {
+        3,           // sizeof fd_table
+        0, nullptr,  // NOTE: no remappings
+        static_cast<uvwasi_size_t>(argc), argv,
+        nullptr,  // TODO: support env
+        0, 1, 2,
+        nullptr,  // NOTE: no special allocator
+    };
+
+    uvwasi_errno_t err = uvwasi_init(&state, &options);
+    if (err != UVWASI_ESUCCESS)
+    {
+        std::cerr << "Failed to initialise UVWASI: " << uvwasi_embedder_err_code_to_string(err)
+                  << "\n";
+        return false;
+    }
+
+    if (!std::filesystem::is_regular_file(argv[0]))
```

Comment:
It maybe better to just relay on failing `wasm_file{argv[0]}` instead of using `<filesystem>` (to avoid building issues with GCC8 and on macOS).

Reply:
Discussed most of that in https://github.com/wasmx/fizzy/pull/329#discussion_r430463086.

`ifstream` is terrible at error reporting, just returns an empty bytestring in case of errors.

Reply:
It is not great, but you can check `wasm_file.fail()` and print that this file cannot be opened (without the reason why).